### ISSUE TITLE
postgres db update requires password setup

### ISSUE THREAD

**Issue by zzh112119** [vsoch/django-nginx-upload#7]
This is a great model and thanks for sharing!

I was trying the cmd line tool ./push.py and faced the following error. There was no useful log emitted from the script. 

```
./push.py /home/workspaces/django-nginx-upload/img/upload.png
PUSH /home/workspaces/django-nginx-upload/img/upload.png
[0 of 0 MB]Expecting value: line 1 column 1 (char 0)
```
Then I moved on to use docker-compose logs nginx shown below and found out a few 5xx errors. Initially I suspected this would be caused by permission or missing folder. But the container showed me all correct configuration. 

```
nginx_1  | 240.10.0.1 - - [14/Apr/2020:04:09:45 +0000] "POST /upload HTTP/1.1" 503 213 "-" "python-requests/2.18.4" "-"
nginx_1  | 2020/04/14 04:12:01 [error] 7#7: *4 failed to create output file "/var/www/images/_upload/4/0000000004" for "upload.png" (2: No such file or directory), client: 240.10.0.1, server: localhost, request: "POST /upload HTTP/1.1", host: "127.0.0.1"
nginx_1  | 240.10.0.1 - - [14/Apr/2020:04:13:07 +0000] "POST /upload HTTP/1.1" 500 162228 "-" "python-requests/2.18.4" "-"
nginx_1  | 2020/04/14 04:13:07 [error] 7#7: *5 failed to remove destination file "/var/www/images/_upload/5/0000000005" after http status 500 (2: No such file or directory) while closing request, client: 240.10.0.1, server: 0.0.0.0:80
nginx_1  | 2020/04/14 04:20:42 [error] 6#6: *1 connect() failed (111: Connection refused) while connecting to upstream, client: 240.10.0.1, server: localhost, request: "POST /upload HTTP/1.1", upstream: "uwsgi://240.10.0.4:3031", host: "127.0.0.1"
nginx_1  | 240.10.0.1 - - [14/Apr/2020:04:20:42 +0000] "POST /upload HTTP/1.1" 502 173 "-" "python-requests/2.18.4" "-"
nginx_1  | 240.10.0.1 - - [14/Apr/2020:04:21:35 +0000] "POST /upload HTTP/1.1" 500 162208 "-" "python-requests/2.18.4" "-"
nginx_1  | 2020/04/14 04:21:35 [error] 6#6: *3 failed to remove destination file "/var/www/images/_upload/2/0000000002" after http status 500 (2: No such file or directory) while closing request, client: 240.10.0.1, server: 0.0.0.0:80
nginx_1  | 240.10.0.1 - - [14/Apr/2020:04:22:05 +0000] "POST /upload HTTP/1.1" 500 162208 "-" "python-requests/2.18.4" "-"
nginx_1  | 2020/04/14 04:22:05 [error] 6#6: *5 failed to remove destination file "/var/www/images/_upload/3/0000000003" after http status 500 (2: No such file or directory) while closing request, client: 240.10.0.1, server: 0.0.0.0:80
nginx_1  | 240.10.0.1 - - [14/Apr/2020:04:24:45 +0000] "POST /upload HTTP/1.1" 500 162208 "-" "python-requests/2.18.4" "-"
nginx_1  | 2020/04/14 04:24:45 [error] 6#6: *7 failed to remove destination file "/var/www/images/_upload/4/0000000004" after http status 500 (2: No such file or directory) while closing request, client: 240.10.0.1, server: 0.0.0.0:80
```
Then I moved on to db's log and found this! This links to a recent change in postgres' docker library here: https://github.com/docker-library/postgres/commit/42ce7437ee68150ee29f5272428aa4fc657dc6dc. 

```
django-nginx-upload git:(master) docker-compose logs db
Attaching to django-nginx-upload_db_1
db_1     | Error: Database is uninitialized and superuser password is not specified.
db_1     |        You must specify POSTGRES_PASSWORD to a non-empty value for the
db_1     |        superuser. For example, "-e POSTGRES_PASSWORD=password" on "docker run".
db_1     | 
db_1     |        You may also use "POSTGRES_HOST_AUTH_METHOD=trust" to allow all
db_1     |        connections without a password. This is *not* recommended.
db_1     | 
db_1     |        See PostgreSQL documentation about "trust":
db_1     |        https://www.postgresql.org/docs/current/auth-trust.html
db_1     | Error: Database is uninitialized and superuser password is not specified.
db_1     |        You must specify POSTGRES_PASSWORD to a non-empty value for the
db_1     |        superuser. For example, "-e POSTGRES_PASSWORD=password" on "docker run".
db_1     | 
db_1     |        You may also use "POSTGRES_HOST_AUTH_METHOD=trust" to allow all
db_1     |        connections without a password. This is *not* recommended.
db_1     | 
db_1     |        See PostgreSQL documentation about "trust":
db_1     |        https://www.postgresql.org/docs/current/auth-trust.html
db_1     | Error: Database is uninitialized and superuser password is not specified.
db_1     |        You must specify POSTGRES_PASSWORD to a non-empty value for the
db_1     |        superuser. For example, "-e POSTGRES_PASSWORD=password" on "docker run".
db_1     | 
db_1     |        You may also use "POSTGRES_HOST_AUTH_METHOD=trust" to allow all
db_1     |        connections without a password. This is *not* recommended.
db_1     | 
db_1     |        See PostgreSQL documentation about "trust":
db_1     |        https://www.postgresql.org/docs/current/auth-trust.html
db_1     | Error: Database is uninitialized and superuser password is not specified.
db_1     |        You must specify POSTGRES_PASSWORD to a non-empty value for the
db_1     |        superuser. For example, "-e POSTGRES_PASSWORD=password" on "docker run".
db_1     | 
db_1     |        You may also use "POSTGRES_HOST_AUTH_METHOD=trust" to allow all
db_1     |        connections without a password. This is *not* recommended.
db_1     | 
db_1     |        See PostgreSQL documentation about "trust":
db_1     |        https://www.postgresql.org/docs/current/auth-trust.html
```

Just to verify this is really the root cause I also grabbed the uwsgi's log.

```
django-nginx-upload git:(master) docker-compose logs uwsgi
Attaching to django-nginx-upload_uwsgi_1
uwsgi_1  | Traceback (most recent call last):
uwsgi_1  |   File "/usr/local/lib/python3.5/site-packages/django/db/backends/base/base.py", line 217, in ensure_connection
uwsgi_1  |     self.connect()
uwsgi_1  |   File "/usr/local/lib/python3.5/site-packages/django/db/backends/base/base.py", line 195, in connect
uwsgi_1  |     self.connection = self.get_new_connection(conn_params)
uwsgi_1  |   File "/usr/local/lib/python3.5/site-packages/django/db/backends/postgresql/base.py", line 178, in get_new_connection
uwsgi_1  |     connection = Database.connect(**conn_params)
uwsgi_1  |   File "/usr/local/lib/python3.5/site-packages/psycopg2/__init__.py", line 127, in connect
uwsgi_1  |     conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
uwsgi_1  | psycopg2.OperationalError: could not connect to server: Connection refused
uwsgi_1  | 	Is the server running on host "db" (240.10.0.2) and accepting
uwsgi_1  | 	TCP/IP connections on port 5432?
uwsgi_1  | 
uwsgi_1  | 
uwsgi_1  | The above exception was the direct cause of the following exception:
uwsgi_1  | 
uwsgi_1  | Traceback (most recent call last):
uwsgi_1  |   File "manage.py", line 15, in <module>
uwsgi_1  |     execute_from_command_line(sys.argv)
uwsgi_1  |   File "/usr/local/lib/python3.5/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line
uwsgi_1  |     utility.execute()
uwsgi_1  |   File "/usr/local/lib/python3.5/site-packages/django/core/management/__init__.py", line 375, in execute
uwsgi_1  |     self.fetch_command(subcommand).run_from_argv(self.argv)
uwsgi_1  |   File "/usr/local/lib/python3.5/site-packages/django/core/management/base.py", line 323, in run_from_argv
uwsgi_1  |     self.execute(*args, **cmd_options)
uwsgi_1  |   File "/usr/local/lib/python3.5/site-packages/django/core/management/base.py", line 364, in execute
uwsgi_1  |     output = self.handle(*args, **options)
uwsgi_1  |   File "/usr/local/lib/python3.5/site-packages/django/core/management/base.py", line 83, in wrapped
uwsgi_1  |     res = handle_func(*args, **kwargs)
uwsgi_1  |   File "/usr/local/lib/python3.5/site-packages/django/core/management/commands/makemigrations.py", line 101, in handle
uwsgi_1  |     loader.check_consistent_history(connection)
uwsgi_1  |   File "/usr/local/lib/python3.5/site-packages/django/db/migrations/loader.py", line 283, in check_consistent_history
uwsgi_1  |     applied = recorder.applied_migrations()
uwsgi_1  |   File "/usr/local/lib/python3.5/site-packages/django/db/migrations/recorder.py", line 73, in applied_migrations
uwsgi_1  |     if self.has_table():
uwsgi_1  |   File "/usr/local/lib/python3.5/site-packages/django/db/migrations/recorder.py", line 56, in has_table
uwsgi_1  |     return self.Migration._meta.db_table in self.connection.introspection.table_names(self.connection.cursor())
uwsgi_1  |   File "/usr/local/lib/python3.5/site-packages/django/db/backends/base/base.py", line 256, in cursor
uwsgi_1  |     return self._cursor()
uwsgi_1  |   File "/usr/local/lib/python3.5/site-packages/django/db/backends/base/base.py", line 233, in _cursor
uwsgi_1  |     self.ensure_connection()
uwsgi_1  |   File "/usr/local/lib/python3.5/site-packages/django/db/backends/base/base.py", line 217, in ensure_connection
uwsgi_1  |     self.connect()
uwsgi_1  |   File "/usr/local/lib/python3.5/site-packages/django/db/utils.py", line 89, in __exit__
uwsgi_1  |     raise dj_exc_value.with_traceback(traceback) from exc_value
uwsgi_1  |   File "/usr/local/lib/python3.5/site-packages/django/db/backends/base/base.py", line 217, in ensure_connection
uwsgi_1  |     self.connect()
uwsgi_1  |   File "/usr/local/lib/python3.5/site-packages/django/db/backends/base/base.py", line 195, in connect
uwsgi_1  |     self.connection = self.get_new_connection(conn_params)
uwsgi_1  |   File "/usr/local/lib/python3.5/site-packages/django/db/backends/postgresql/base.py", line 178, in get_new_connection
uwsgi_1  |     connection = Database.connect(**conn_params)
uwsgi_1  |   File "/usr/local/lib/python3.5/site-packages/psycopg2/__init__.py", line 127, in connect
uwsgi_1  |     conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
uwsgi_1  | django.db.utils.OperationalError: could not connect to server: Connection refused
uwsgi_1  | 	Is the server running on host "db" (240.10.0.2) and accepting
uwsgi_1  | 	TCP/IP connections on port 5432?
```

According to the log from the db, I tried to modify docker-compose file to: 

```
db:
  image: postgres
  environment:
    - POSTGRES_HOST_AUTH_METHOD=trust
```

However, this is still outputting errors to me. It sounds like we have to setup the password for authentication somehow.

```
docker-compose logs db
Attaching to django-nginx-upload_db_1
db_1     | 
db_1     | PostgreSQL Database directory appears to contain a database; Skipping initialization
db_1     | 
db_1     | 2020-04-14 05:15:11.639 UTC [1] LOG:  starting PostgreSQL 12.2 (Debian 12.2-2.pgdg100+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 8.3.0-6) 8.3.0, 64-bit
db_1     | 2020-04-14 05:15:11.640 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
db_1     | 2020-04-14 05:15:11.640 UTC [1] LOG:  listening on IPv6 address "::", port 5432
db_1     | 2020-04-14 05:15:11.664 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
db_1     | 2020-04-14 05:15:11.700 UTC [27] LOG:  database system was shut down at 2020-04-14 05:13:54 UTC
db_1     | 2020-04-14 05:15:11.712 UTC [1] LOG:  database system is ready to accept connections
db_1     | 2020-04-14 05:15:12.863 UTC [34] FATAL:  password authentication failed for user "postgres"
db_1     | 2020-04-14 05:15:12.863 UTC [34] DETAIL:  Role "postgres" does not exist.
db_1     | 	Connection matched pg_hba.conf line 95: "host all all all md5"
db_1     | 2020-04-14 05:15:13.261 UTC [35] FATAL:  password authentication failed for user "postgres"
db_1     | 2020-04-14 05:15:13.261 UTC [35] DETAIL:  Role "postgres" does not exist.
db_1     | 	Connection matched pg_hba.conf line 95: "host all all all md5"
db_1     | 2020-04-14 05:15:13.642 UTC [36] FATAL:  password authentication failed for user "postgres"
db_1     | 2020-04-14 05:15:13.642 UTC [36] DETAIL:  Role "postgres" does not exist.
db_1     | 	Connection matched pg_hba.conf line 95: "host all all all md5"
db_1     | 2020-04-14 05:15:14.028 UTC [37] FATAL:  password authentication failed for user "postgres"
db_1     | 2020-04-14 05:15:14.028 UTC [37] DETAIL:  Role "postgres" does not exist.
```

```
docker-compose logs uwsgi
uwsgi_1  | [pid: 19|app: 0|req: 1/1] 240.10.0.1 () {38 vars in 583 bytes} [Tue Apr 14 05:15:15 2020] POST /upload => generated 161006 bytes in 97 msecs (HTTP/1.1 500) 4 headers in 130 bytes (2 switches on core 0)
uwsgi_1  | <WSGIRequest: POST '/complete/'>
uwsgi_1  | Internal Server Error: /complete/
uwsgi_1  | Traceback (most recent call last):
uwsgi_1  |   File "/usr/local/lib/python3.5/site-packages/django/db/backends/base/base.py", line 217, in ensure_connection
uwsgi_1  |     self.connect()
uwsgi_1  |   File "/usr/local/lib/python3.5/site-packages/django/db/backends/base/base.py", line 195, in connect
uwsgi_1  |     self.connection = self.get_new_connection(conn_params)
uwsgi_1  |   File "/usr/local/lib/python3.5/site-packages/django/db/backends/postgresql/base.py", line 178, in get_new_connection
uwsgi_1  |     connection = Database.connect(**conn_params)
uwsgi_1  |   File "/usr/local/lib/python3.5/site-packages/psycopg2/__init__.py", line 127, in connect
uwsgi_1  |     conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
uwsgi_1  | psycopg2.OperationalError: fe_sendauth: no password supplied
```

```
docker-compose logs nginx
Attaching to django-nginx-upload_nginx_1
nginx_1  | 2020/04/14 05:15:15 [warn] 7#7: *1 an upstream response is buffered to a temporary file /var/cache/nginx/uwsgi_temp/2/00/0000000002 while reading upstream, client: 240.10.0.1, server: localhost, request: "POST /upload HTTP/1.1", upstream: "uwsgi://240.10.0.3:3031", host: "127.0.0.1"
nginx_1  | 240.10.0.1 - - [14/Apr/2020:05:15:15 +0000] "POST /upload HTTP/1.1" 500 161006 "-" "python-requests/2.18.4" "-"
nginx_1  | 2020/04/14 05:15:15 [error] 7#7: *1 failed to remove destination file "/var/www/images/_upload/1/0000000001" after http status 500 (2: No such file or directory) while closing request, client: 240.10.0.1, server: 0.0.0.0:80
nginx_1  | 240.10.0.1 - - [14/Apr/2020:05:15:18 +0000] "POST /upload HTTP/1.1" 500 161006 "-" "python-requests/2.18.4" "-"
nginx_1  | 2020/04/14 05:15:18 [error] 7#7: *3 failed to remove destination file "/var/www/images/_upload/3/0000000003" after http status 500 (2: No such file or directory) while closing request, client: 240.10.0.1, server: 0.0.0.0:80

```
This is the furthest I've got today. Please let me know if I'm missing anything here and if there's a good way to fix the issue. 

Thanks!

**Comment by vsoch** [comment]
You were very close! You needed to add the environment variable as you did, but then stop, remove, and rebuild containers.

```bash
docker-compose stop
docker-compose rm
docker-compose build
```
And bring them back up 

```bash
docker-compose up -d
```

Your upload issue is a part of the README that you must have missed - since the upload folders are created in the container but bound over via volume, you have to create them from the host:
https://github.com/vsoch/django-nginx-upload#usage. Here is what you see without doing that, and the "folder not found" is logical - there are no subfolders to upload to!

```bash
/ # ls -l /var/www/images/_upload
total 0
```
Copy pasting the command from nginx/Dockerfile (or the README instruction from the outside) fixes the issue.

```sh
 # ls /var/www/images/_upload/
1       2       3       4       5       6       7       8       9
```

Then we can check the interface upload first

![image](https://user-images.githubusercontent.com/814322/79199603-75322680-7df2-11ea-951c-f7429e46cbf7.png)

And then command line

```bash
$ python push.py img/upload.png 
PUSH /home/vanessa/Documents/django-nginx-upload/img/upload.png
[0 of 0 MB]
[Return status 200 Upload Complete]
```
So TLDR: The issue comes down to binding the folders on the host, since we bind /var/www/images
the folders in the container are essentially wiped, you need to create the paths. I've just pushed a fix for the docker-compose file, let me know if you run into any other trouble!

**Comment by zzh112119** [comment]
Thanks for the update! I totally missed the mkdir part and was able to correct the issue by using that command. The upload was successful on both GUI interface and the push script.

There's a totally unrelated question to this issue I'd like to ask for help. Have you seen 301 response before when you push/POST a file to the nginx server? I'm trying to follow what you have here and the nginx upload module example to create my own package but getting 301 response when trying to use the ./push.py or simply `curl -v -F file=@fakefile http://localhost:80/upload`. I initially thought it would be a user permission issue or a network configuration issue but couldn't find the right path to figure it out...

The symptom is that the response is logged as below. And the file is never uploaded to the container/volume.
```
curl -v -F "file=@./img/upload.png" http://localhost:80/upload
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 80 (#0)
> POST /upload HTTP/1.1
> Host: localhost
> User-Agent: curl/7.58.0
> Accept: */*
> Content-Length: 20439
> Content-Type: multipart/form-data; boundary=------------------------daca1eac665c7b02
> Expect: 100-continue
> 
< HTTP/1.1 301 Moved Permanently
< Server: nginx/1.16.1
< Date: Wed, 15 Apr 2020 01:12:42 GMT
< Content-Type: text/html
< Content-Length: 169
< Location: http://localhost/upload/
< Connection: keep-alive
* HTTP error before end of send, stop sending
< 
<html>
<head><title>301 Moved Permanently</title></head>
<body>
<center><h1>301 Moved Permanently</h1></center>
<hr><center>nginx/1.16.1</center>
</body>
</html>

```

**Comment by vsoch** [comment]
It’s concerning that the port is removed (notice the 80 is gone) do you have another something / server running on localhost?

**Comment by vsoch** [comment]
The example here is really just a starter setup - you likely need a more advanced Nginx.conf for more realistic use cases. You can look at https://github.com/singularityhub/sregistry for a more realistic one, although it’s going away soon in favor of using Minio.

**Comment by zzh112119** [comment]
Yep will look more into the port configuration. Thanks for the link to the repo. I also saw your blog on the MinIO multipart upload (this one https://vsoch.github.io/2020/s3-minio-multipart-presigned-upload/). It is very interesting to see minio is able to provide s3 like functionality. LOTS of appreciation for the effort you've made for these examples.
